execution/executor/src/chunk_executor.rs

```diff
@@ -35,12 +35,113 @@ use storage_interface::sync_proof_fetcher::SyncProofFetcher;
 use storage_interface::{cached_state_view::CachedStateView, DbReaderWriter, ExecutedTrees};
 
 pub struct ChunkExecutor<V> {
+    db: DbReaderWriter,
+    inner: RwLock<Option<ChunkExecutorInner<V>>>,
+}
+
+impl<V: VMExecutor> ChunkExecutor<V> {
+    pub fn new(db: DbReaderWriter) -> Self {
+        Self {
+            db,
+            inner: RwLock::new(None),
+        }
+    }
+
+    fn maybe_initialize(&self) -> Result<()> {
+        if self.inner.read().is_none() {
+            self.reset()?;
+        }
+        Ok(())
+    }
+}
+
+impl<V: VMExecutor> ChunkExecutorTrait for ChunkExecutor<V> {
+    fn execute_chunk(
+        &self,
+        txn_list_with_proof: TransactionListWithProof,
+        verified_target_li: &LedgerInfoWithSignatures,
+        epoch_change_li: Option<&LedgerInfoWithSignatures>,
+    ) -> Result<()> {
+        self.maybe_initialize()?;
+        self.inner
+            .read()
+            .as_ref()
+            .expect("not reset")
+            .execute_chunk(txn_list_with_proof, verified_target_li, epoch_change_li)
+    }
+
+    fn apply_chunk(
+        &self,
+        txn_output_list_with_proof: TransactionOutputListWithProof,
+        verified_target_li: &LedgerInfoWithSignatures,
+        epoch_change_li: Option<&LedgerInfoWithSignatures>,
+    ) -> Result<()> {
+        self.inner.read().as_ref().expect("not reset").apply_chunk(
+            txn_output_list_with_proof,
+            verified_target_li,
+            epoch_change_li,
+        )
+    }
+
+    fn commit_chunk(&self) -> Result<ChunkCommitNotification> {
+        self.inner
+            .read()
+            .as_ref()
+            .expect("not reset")
+            .commit_chunk()
+    }
+
+    fn execute_and_commit_chunk(
+        &self,
+        txn_list_with_proof: TransactionListWithProof,
+        verified_target_li: &LedgerInfoWithSignatures,
+        epoch_change_li: Option<&LedgerInfoWithSignatures>,
+    ) -> Result<ChunkCommitNotification> {
+        // Re-sync with DB, make sure the queue is empty.
+        self.reset()?;
+        self.inner
+            .read()
+            .as_ref()
+            .expect("not reset")
+            .execute_and_commit_chunk(txn_list_with_proof, verified_target_li, epoch_change_li)
+    }
+
+    fn apply_and_commit_chunk(
+        &self,
+        txn_output_list_with_proof: TransactionOutputListWithProof,
+        verified_target_li: &LedgerInfoWithSignatures,
+        epoch_change_li: Option<&LedgerInfoWithSignatures>,
+    ) -> Result<ChunkCommitNotification> {
+        // Re-sync with DB, make sure the queue is empty.
+        self.reset()?;
+        self.inner
```

Comment:
nit: can you use a helper to avoid repeating this pattern in every API call?

Reply:
haha, how to do that? Wrapping this `self.inner.read().as_ref().expect("not reset")` does not work...